tools/chpldoc/findUndocumentedSymbols.py

```diff
@@ -0,0 +1,387 @@
+
+
+def get_node_name(node: dyno.AstNode) -> str:
+    """
+    get the name of a node, even when it may not have a `name()` method
+
+    if this is a member of a aggregate type, also gets that name
+    if this is a member of a module, also gets that name
+
+    For a multidecl, gets all the names
+
+    """
+
+    def get_simple_node_name(node: dyno.AstNode) -> str:
+        name: str = None
+        if hasattr(node, "name"):
+            name = node.name()
+        elif isinstance(node, dyno.MultiDecl):
+            name = "MultiDecl[" + ", ".join(c.name() for c in node) + "]"
+        else:
+            name = str(node)
+        return name
+
+    name = get_simple_node_name(node)
+
+    # handles secondary methods
+    if hasattr(node, "this_formal") and (this := node.this_formal()) and (typename := this.type_expression()):
+        aggregate_name = get_simple_node_name(typename)
+        name = f"{aggregate_name}.{name}"
+    # handles primary methods and fields
+    elif (p := node.parent()) and (isinstance(p, dyno.AggregateDecl) or isinstance(p, dyno.Interface)):
+        aggregate_name = get_simple_node_name(p)
+        name = f"{aggregate_name}.{name}"
+
+    if (mod := get_module(node.parent())) and mod.kind() != "implicit":
+        module_name = mod.name()
+        name = f"{module_name}.{name}"
+
+    return name
+
+
+def lookAhead(iterable, window=1):
+    items, nexts = itertools.tee(iterable, 2)
+    nexts = itertools.islice(nexts, window, None)
+    return itertools.zip_longest(items, nexts)
+
+
+"""
+main code
+"""
+
+
+class FindUndocumentedSymbols:
+    def __init__(
+        self,
+        ast: List[dyno.AstNode],
+        ignore_deprecated: bool = False,
+        ignore_unstable: bool = False,
+    ):
+        self.ast = ast
+        self.ignore_deprecated = ignore_deprecated
+        self.ignore_unstable = ignore_unstable
+
+    # (pattern, extra check fun)
+    documentable_symbol_patterns = {
+        "function": (
+            dyno.Function,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl, dyno.Interface)
+            ),
+        ),
+        "module": (dyno.Module, lambda node, match: node.kind() != "implicit"),
+        "aggregate_decl": (
+            dyno.AggregateDecl,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "decl": (
+            dyno.Variable,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "multi_decl": (
+            dyno.MultiDecl,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
+        ),
+        "enum": (
+            dyno.Enum,
+            lambda node, match: isinstance(
+                node.parent(), (dyno.Module, dyno.AggregateDecl)
+            ),
```

Comment:
Consider extracting the extra check funcs into `def`'ed functions and re-using them.